File: Jim/src/sk/fiit/jim/agent/highskill/Kick.java
package sk.fiit.jim.agent.highskill;

import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.EnvironmentModel;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Vector3D;

/**
 *
 * Reimplemented from Ruby to Java by roman moravcik
 */
public class Kick extends HighSkill {

    private Vector3D kickTarget;

    public Kick(Vector3D kickTarget) {
        this.kickTarget = kickTarget;
    }

    @Override
    public LowSkill pickLowSkill() {
        AgentInfo agentInfo = AgentInfo.getInstance();
        AgentModel agentModel = AgentModel.getInstance();
       
        Vector3D ballPosition = agentInfo.ballControlPosition();

        if (EnvironmentModel.beamablePlayMode() && !EnvironmentModel.isKickOffPlayMode()) {
            agentInfo.loguj("beam");
            return null;
        } else if (agentModel.isOnGround()) {
            agentInfo.loguj("fall");
            return null;
        } else if (ballUnseen() > 3) {
            agentInfo.loguj("ball unseen");
            return null;
        } else if (Math.abs(ballPosition.getY()) > 0.7) {
            agentInfo.loguj("very big y");
            return null;
        } else if (Math.abs(ballPosition.getX()) > 0.7) {
            agentInfo.loguj("very big y");
            return null;
        } else if (ballPosition.getY() < 0.7 && ballPosition.getY() > 0.25) {
            agentInfo.loguj("go a bit");
            return LowSkills.get("walk_slow2");
        } else if (ballPosition.getX() > 0.1 && ballPosition.getX() < 0.7) {
            agentInfo.loguj("big x");
            return LowSkills.get("step_right");
        } else if (ballPosition.getX() < -0.1 && ballPosition.getX() > -0.7) {
            agentInfo.loguj("small x");
            return LowSkills.get("step_left");
        } else if (ballPosition.getX() > 0.0) {
            agentInfo.loguj("kick right");
            return getKick("right");
        } else if (ballPosition.getX() < 0.0) {
            agentInfo.loguj("kick left");
            return getKick("left");
        } else {
            agentInfo.calculateDistance(ballPosition, ballPosition);
            agentInfo.loguj("unknown kick distance");
            return null;
        }
    }

    private LowSkill getKick(String leg) {
        
        // FIXME: calculateDistance by mala byt static z Vector3D.
        // FIXME: druhy parameter je len odhad kedze v skripte ta premmenna ani nebola inicializovana.
        Vector3D ballPosition = WorldModel.getInstance().getBall().getPosition();
        double kickDistance = AgentInfo.getInstance().calculateDistance(ballPosition, kickTarget);

        if (kickDistance > 4 && leg == "right") {
            System.out.println("faster");
            return LowSkills.get("kick_step_strong_right");
            // return LowSkills.get("kick_right_faster");
        } 
        else if (kickDistance > 4 && leg == "left") {
            System.out.println("faster");
            return LowSkills.get("kick_step_strong_left");
            // return LowSkills.get("kick_left_faster");
        } 
        else if (kickDistance <= 4 && kickDistance > 1.5 && leg == "right") {
            System.out.println("fast");
            return LowSkills.get("kick_right_faster");
            // return LowSkills.get("kick_right_normal");
        } 
        else if (kickDistance <= 2.5 && kickDistance > 1.5 && leg == "left") {
            System.out.println("fast");
            return LowSkills.get("kick_left_faster");
            // return LowSkills.get("kick_left_normal");
        } 
        else if (kickDistance <= 1.5 && kickDistance > 0.5 && leg == "right") {
            System.out.println("normal");
            return LowSkills.get("kick_right_normal");
            // return LowSkills.get("kick_right_slow");
        } 
        else if (kickDistance <= 1.5 && kickDistance > 0.5 && leg == "left") {
            System.out.println("normal");
            return LowSkills.get("kick_left_normal");
            // return LowSkills.get("kick_left_slow");
        } 
        else {
            return null;
        }
    }

    @Override
    public void checkProgress() {
    }

    // FIXME: to iste je aj v Plan.java, treba dat do dakych spolocnych konstant.
    public static double ballUnseen() {
        return EnvironmentModel.SIMULATION_TIME - WorldModel.getInstance().getBall().getLastTimeSeen();
    }
}

File: Old_xml_and_scripts/Jim/high_skills/new-18.5.2013/kick.rb
require 'high_skills/ruby_high_skill.rb'

include_class Java::sk.fiit.robocup.library.geometry.Vector3D

class Kick< RubyHighSkill

	
	def initialize (kick_target)
		super()
		@agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
    	@agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance
    	@mathModel = Java::sk.fiit.robocup.library.geometry.Vector3D.clone 
    	@worldModel = Java::sk.fiit.jim.agent.models.WorldModel.getInstance   
		
		@kick_target = kick_target
	end

  def pickLowSkill
	
	
    
    ball_pos = @agentInfo.ballControlPosition()
   
    
        
    puts ball_pos.getY()
    puts ball_pos.getX()
    
    case
      when (EnvironmentModel.beamablePlayMode and not EnvironmentModel.isKickOffPlayMode) :
        puts "beam" 
        return nil
      when @agentModel.falled? :
        puts "fall"
        return nil
      when ball_unseen > 3 :
        puts "ball unseen"
        return nil

      when ball_pos.getY().abs > 0.7 
        puts "very big Y"
        return nil    
        when ball_pos.getX().abs > 0.7
              puts "very big X"
              return nil
                  
	  when (ball_pos.getY() < 0.7 and ball_pos.getY() > 0.25) 
        puts "go a bit"
        return get_skill("walk_slow2")
    
   
      when (ball_pos.getX() > 0.1 and ball_pos.getX() < 0.7)
        puts "big X"
        return get_skill("step_right")
      when (ball_pos.getX() < -0.1 and ball_pos.getX() > -0.7)
        puts "small X"
        return get_skill("step_left")
    
      when (ball_pos.getX() > 0.0) 
        puts "kick right"
        @ballR=ball_pos.getR()        
        kick_dist "right", ball_pos
      when (ball_pos.getX() < 0.0) 
        puts "kick left"        
        @ballR=ball_pos.getR()
        kick_dist "left", ball_pos
     
      else
        puts "wtf????"
        return nil
      end

  end

  def kick_dist leg, ball_pos
  	
 	
 	 @kick_target
 	 ball_pos = @worldModel.getBall().getPosition
    
   	@kick_target_dist = @agentInfo.calculateDistance(ball_pos,@kick_target)
 	
 	
  	if(@kick_target_dist > 2.5 and leg == "right")
  		puts "faster"
  		get_skill("kick_step_strong_right")
  		#get_skill("kick_right_faster")
	elsif (@kick_target_dist > 2.5 and leg=="left")
 		puts "faster"
 		get_skill("kick_step_strong_left")
 		#get_skill("kick_left_faster")
 	
 	elsif(@kick_target_dist <= 2.5 and @kick_target_dist > 1.5  and leg=="right")
  		puts "fast"
  		get_skill("kick_right_faster")
  		#get_skill("kick_right_normal")
	elsif(@kick_target_dist <= 2.5 and @kick_target_dist > 1.5  and leg=="left")
		puts "fast"
 		get_skill("kick_left_faster")
 		#get_skill("kick_left_normal")
 	elsif(@kick_target_dist <= 1.5 and @kick_target_dist > 0.5 and leg=="right")
  		puts "normal"
  		get_skill("kick_right_normal")
  		#get_skill("kick_right_slow")
	elsif (@kick_target_dist <= 1.5 and @kick_target_dist > 0.5 and leg=="left")
 		puts "normal"
 		get_skill("kick_left_normal")
 		#get_skill("kick_left_slow")
 	else
 		return nil
  	end
  
  end 

  
  def checkProgress
  #  raise "Fallen" if me.on_ground?
  end
end
#Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(0, 0, 0)
#in order not to defer the startup cost
Kick.new (@kick_target)
